datafusion/sql/src/unparser/plan.rs

```diff
@@ -176,7 +250,9 @@ impl Unparser<'_> {
                 )
             }
             LogicalPlan::Aggregate(_agg) => {
```

Comment:
I'm currently assuming all aggregate plan nodes are preceded by a projection plan node. These two nodes are then handled together simultaneously in the above projection node block. 

@alamb I was wondering if you know:

1. If an Aggregate plan node can ever NOT be preceded by a projection plan node?
2. If so, under what circumstances and what does that mean intuitively?

Reply:
This does indeed become more tricky. I guess the schema computed at every layer of the plan passes up enough information for execution but maybe not for our Unparser use-case. The specific columns being selected are no longer needed/known at the aggregation step.

I wonder if it's worth creating a structure next to the plan that builds up schema-like data for serialization purpose.

Reply:
@devinjdangelo 
I have done a lot of work on this problem independently - although I think this approach of trying to rebuild the sqlparser-rs AST and then using it to generate SQL is nicer.

Regarding your questions you start to see all kinds of crazy plans when you start trying to unwind aggregations:

```sql
// Projection: pc.contacttypeid, ctypename, nocontacts
//   Sort: COUNT(*) AS nocontacts DESC NULLS FIRST
//     Projection: pc.contacttypeid, pc.name AS ctypename, COUNT(*) AS nocontacts, COUNT(*)
//       Filter: COUNT(*) >= Int64(100)
//         Aggregate: groupBy=[[pc.contacttypeid, pc.name]], aggr=[[COUNT(*)]]
```

Reply:
Thank you @seddonm1! You are right. I cooked up a test case that looks similar to the plan you show.

```
Sort: COUNT(*) ASC NULLS LAST
  Projection: person.id, COUNT(*), person.first_name
    Filter: COUNT(*) > Int64(5)
      Aggregate: groupBy=[[person.first_name, person.id]], aggr=[[COUNT(*)]]
        TableScan: person
```

```sql
select id, count(*), first_name 
from person 
group by first_name, id 
having count(*)>5 
order by count(*)
``` 

This case will of course fail in the PR. I think that is OK, as we can work on support for more complex aggregations and additional edge cases we find in future PRs.

Reply:
It will get interesting as we will need to start writing logic that peaks up and down the plan :thinking:

Reply:
I can put my current code on a gist if that would help. i had to do multiple nested matching conditions to be able to traverse and extract nested data.

What would be ideal is if this could drive changes in the LogicalPlan nodes to help `unparse` them.

Reply:
That would be great! Also feel free to open a PR to DataFusion if you are interested in collaborating on this functionality.